src/pybind/mgr/cephadm/services/monitoring.py

```diff
@@ -87,6 +70,44 @@ def generate_config(self, daemon_spec: CephadmDaemonDeploySpec) -> Tuple[Dict[st
         }
         return config_file, sorted(deps)
 
+    def prepare_certificates(self) -> Tuple[str, str]:
+        cert = self.mgr.get_store('grafana_crt')
+        pkey = self.mgr.get_store('grafana_key')
+        (org, cn) = (None, None)
+        if cert and pkey:
+            try:
+                (org, cn) = get_cert_issuer_info(cert)
+                verify_tls(cert, pkey)
+            except ServerConfigException as e:
+                logger.warning(f'Provided grafana TLS certificates are invalid: {e}')
+                cert, pkey = None, None
+
+        # we only re-generate self signed certificates if the failed ones
+        # were originaly generated by cephadm or they are empty
+        if not (cert and pkey) \
+           and ((org == 'Ceph' and cn == 'cephadm') or (org is None and cn is None)):
+            cert, pkey = create_self_signed_cert('Ceph', 'cephadm')
+            self.mgr.set_store('grafana_crt', cert)
+            self.mgr.set_store('grafana_key', pkey)
+            if 'dashboard' in self.mgr.get('mgr_map')['modules']:
+                self.mgr.check_mon_command({
+                    'prefix': 'dashboard set-grafana-api-ssl-verify',
+                    'value': 'false',
+                })
+            self.mgr.remove_health_warning('GRAFANA_CERT_ERROR')  # clear if any
+        else:
+            # the certificate was not generated by cephadm, we cannot overwrite
+            # it by new self-signed ones. Let's warn the user to fix the issue
+            err_msg = """
+            Detected invalid grafana certificates. Set mgr/cephadm/grafana_crt
+            and mgr/cephadm/grafana_key to valid certificates or run:
+               > ceph orch daemon redeploy <grafana-daemon>
```

Comment:
If I'm not mistaken, now that we don't overwrite if the cert/key are used generated, won't this redeploy command just end up back here making this same warning? Should we instead be telling the user, if they don't want to provide working certs themselves, to remove the config options so that we'll know to make our own certs the next time around (which seems to be what you're already doing in the other error message in `_check_certificates`)?

Reply:
Yes, you are right. I think the message used in `_check_certificates` is more clear and provide direct instructions on how to proceed. I'll use the same message here as well.